### DISCUSSION TITLE
Recommended Library for algorithms to use with hipSYCL?

### DISCUSSION THREAD

**Discussion by BenBrock** [AdaptiveCpp/AdaptiveCpp#813] [Q&A] [answered]
Is there a recommended library to use for C++ algorithms (reduce, partial_sum, etc.) within hipSYCL?  I'm thinking along the lines of [oneDPL](https://github.com/oneapi-src/oneDPL).

I would also be happy to use oneDPL directly, if it's possible to use it within hipSYCL.  However, there seem to be issues with oneAPI-specific features, and I haven't been able to sort them out myself.

**Comment by illuhad** [comment, marked as answer]
Hi!

We have tried getting oneDPL to work with hipSYCL, and managed to get large portions of it working: https://github.com/hipsycl/onedpl
However, this is based on an old oneDPL version. It has turned out that it is not really maintainable with the resources we have unless the oneDPL devs put hipSYCL in CI and validate with it as well instead of just looking at DPC++. Unforuntantely, we don't have the resources to participate in the oneDPL development to the extent that would be required to pull this off.

Of course, in principle any library written in conformant SYCL 2020 should be able to work with any conformant SYCL 2020 implementation. Unfortunately, such a library does not exist yet, and no implementation currently is fully SYCL 2020 conformant.

If you only need a select few algorithms, what I would recommend is to just rely on hipSYCL's source code interoperability features and hook into vendor libraries like thrust, CUB, rocPRIM etc. Since hipSYCL even allows you to mix-and-match SYCL kernel code with CUDA or HIP code, you can even use such libraries inside your SYCL kernels, e.g. for work group operations.
This will most likely also result in a better performance than oneDPL.

**Reply by BenBrock** [reply]
Thanks for the quick answer. I will take a deeper look at oneDPL, as well as get in touch with the oneDPL people to see if it's possible to fix the issues causing incompatibility with pure SYCL compilers. It would be nice if there were an implementation of STL algorithms we could use with all compilers.

If it's not possible to get oneDPL working, I will likely do as you suggest and implement a shim layer that calls either oneDPL or other vendor STL algorithm implementations. Perhaps I could develop this publicly, as I imagine this would be useful to everyone working with SYCL.

I also saw [an implementation](https://github.com/KhronosGroup/SyclParallelSTL) from the Khronos Group, but it appears to not have been maintained for quite some time. Do you have any impressions as to whether it is currently usable?

**Reply by illuhad** [reply]
Alright. Please let me know if you hit any issues during your endeavour. It might be more complex than just solving compilation issues (which is certainly required anyway), for example oneDPL might also select execution parameters (e.g. work group sizes etc) in a way that doesn't make sense with hipSYCL, potentially leading to inefficient execution.

I believe a lot of people might be interested in such a shim layer for vendor libraries if it is open source and publicly available :-) As I've said, performance-wise, this might be your best bet.

As you say, the SYCL parallel STL has not been maintained for some time. I've not tried it for quite a while, but it used to work with hipSYCL. It targets the older SYCL 1.2.1 standard which - for the most part - is a subset of SYCL 2020, so in principle it should still work, unless you use hierarchical parallelism which is one of the problematic aspects of SYCL 1.2.1. Also, it might not be as convenient as SYCL 2020 for the user, since SYCL 1.2.1 mandates that the user provides explicit kernel names. As far as I remember, SYCL parallel STL also enforces this.
With that said, I'm sure people won't reject pull requests if you want to update it :-)

**Comment by StarGazerM** [comment]
seem its not easy to compile thrust code in acpp, these library will check macro for __CUDACC__ to see if using nvcc. Even if we can force bypass macro,  cuda_runtime is pretty bad organized for namespace and will conflict with many standard c library, for example `abs`

**Reply by illuhad** [reply]
Why would you want to compile thrust with acpp?

AdaptiveCpp has its own algorithm library:  https://github.com/AdaptiveCpp/AdaptiveCpp/blob/develop/doc/algorithms.md

**Reply by StarGazerM** [reply]
ahh, I am mostly using s algorithm from AdaptiveCpp but only one cavet is  I need a radix sort, (I might ends up just write my own), so I look into  how to use thrust together. I understand that  I can just use CFFI to link cuda code, but just want to explore if there is option to keep my library head-only.

**Reply by illuhad** [reply]
If you do end up writing one, please consider contributing it :)
